lib/rom/sql/attribute.rb

```diff
@@ -14,6 +14,28 @@ class Attribute < ROM::Schema::Attribute
       # Error raised when an attribute cannot be qualified
       QualifyError = Class.new(StandardError)
 
+      TypeExtensions = {}
+
+      # Registers a set of operations supported for a specific type
+      #
+      # @example
+      #   ROM::SQL::Attribute.register_type(ROM::SQL::Types::PG::JSONB) do
+      #     def contains(type, keys)
+      #       Sequel::Postgres::JSONBOp.new(type.meta[:name]).contains(keys)
+      #     end
+      #   end
+      #
+      # @param [Dry::Types::Type] type Type
+      #
+      # @api public
+      def self.register_type(type, &block)
+        raise ArgumentError, "Type #{ type } already registered" if TypeExtensions.key?(type)
+        mod = Module.new(&block)
+        ctx = Object.new.extend(mod)
+        functions = mod.instance_methods.each_with_object({}) { |m, ms| ms[m] = ctx.method(m) }
+        TypeExtensions[type] = functions
```

Comment:
I think it's smart, in a good way. After all we need extensions per type, not for all attribute objects. One concern is that we may have name clashes with sequel's sql expressions, that we delegate to. Maybe a simple check if a given method name doesn't clash would be a good idea.